source/common/network/utility.cc

```diff
@@ -249,8 +249,13 @@ bool Utility::isLocalConnection(const Network::ConnectionSocket& socket) {
   // while the client is not actually local. Example could be an iptables intercepted
   // connection. However, this is a rare exception and such assumption results in big
   // performance optimization.
-  if (remote_address->type() == Envoy::Network::Address::Type::Pipe ||
-      remote_address == socket.localAddress() || isLoopbackAddress(*remote_address)) {
+  if (remote_address->type() != Envoy::Network::Address::Type::Ip ||
+      isLoopbackAddress(*remote_address) ||
+      (remote_address->ip()->version() == Address::IpVersion::v4
+           ? remote_address->ip()->ipv4()->address() ==
+                 socket.localAddress()->ip()->ipv4()->address()
+           : remote_address->ip()->ipv6()->address() ==
+                 socket.localAddress()->ip()->ipv6()->address())) {
```

Comment:
This is a lot to read in a conditional statement. Is there a way you can add the needed comparison into an equality operator on the appropriate object to make this statement look a lot clearer? Then you can hide the complexity of this and v4 vs. v6 in the right place.

Reply:
Envoy does not have a generic address type for an IP address without a port number. Typically an IP address type with a zero port number is used to approximate it, but real connection addresses always have non-zero port numbers, so given the current type system this is the only  way to implement the comparison here. We could move this to an utility function (e.g., `isSameIPAddress(addr1, addr2)`) to make it easier to read, though.

Reply:
Ah, that's right. I think it would be best to use a utility function, then.

Reply:
Done.